src/twisted/trial/_dist/workerreporter.py

```diff
@@ -56,48 +215,85 @@ def _getFrames(self, failure: Failure) -> List[str]:
             frames.extend([frame[0], frame[1], str(frame[2])])
         return frames
 
-    def addSuccess(self, test):
+    def _call(self, f: Callable[[], T]) -> None:
         """
-        Send a success over.
+        Call L{f} if and only if a "result reporting" context is active.
+
+        @param f: A function to call.  Its result is accumulated into the
+            result reporting context.  It may return a L{Deferred} or a
+            coroutine or synchronously raise an exception or return a result
+            value.
+
+        @raise ValueError: If no result reporting context is active.
+        """
+        if self._reporting is not None:
+            self._reporting.record(maybeDeferred(f))
+        else:
+            raise ValueError(
+                "Cannot call command outside of reporting context manager."
+            )
+
+    def addSuccess(self, test: PyUnitTestCase) -> None:
+        """
+        Send a success to the parent process.
+
+        This must be called in context managed by L{gatherReportingResults}.
         """
         super().addSuccess(test)
         testName = test.id()
-        self.ampProtocol.callRemote(managercommands.AddSuccess, testName=testName)
+        self._call(
+            lambda: self.ampProtocol.callRemote(  # type: ignore[no-any-return]
+                managercommands.AddSuccess, testName=testName
+            )
+        )
 
-    def addError(self, test, error):
+    async def _addErrorFallible(self, testName: str, errorObj: TrialFailure) -> None:
```

Comment:
Maybe make it public.

The docs are good enough and I hope people will not abuse it.

```suggestion
    async def addErrorFallible(self, testName: str, errorObj: TrialFailure) -> None:
```